File: reports.py
"""Report query builders for PowerFlow."""

from database import db, rows_to_list


def _date_clause(column, date_from, date_to):
    clauses = []
    params = []
    if date_from:
        clauses.append(f"{column} >= ?")
        params.append(date_from)
    if date_to:
        clauses.append(f"{column} <= ?")
        params.append(date_to)
    where = f" AND {' AND '.join(clauses)}" if clauses else ""
    return where, params


def build_full_report(date_from=None, date_to=None):
    inv_where, inv_params = _date_clause("inv.invoice_date", date_from, date_to)
    po_where, po_params = _date_clause("po.order_date", date_from, date_to)
    so_where, so_params = _date_clause("so.order_date", date_from, date_to)
    qt_where, qt_params = _date_clause("q.quote_date", date_from, date_to)

    with db() as conn:
        # --- Financial KPIs ---
        revenue = conn.execute(
            f"""
            SELECT COALESCE(SUM(total), 0) AS total,
                   COALESCE(SUM(amount_paid), 0) AS collected,
                   COALESCE(SUM(total - amount_paid), 0) AS outstanding,
                   COUNT(*) AS invoice_count
            FROM invoices inv
            WHERE inv.status NOT IN ('void', 'draft') {inv_where}
            """,
            inv_params,
        ).fetchone()

        cogs = conn.execute(
            f"""
            SELECT COALESCE(SUM(pol.line_total), 0) AS total
            FROM purchase_order_lines pol
            JOIN purchase_orders po ON po.id = pol.purchase_order_id
            WHERE po.status IN ('received', 'partial') {po_where}
            """,
            po_params,
        ).fetchone()

        inventory_cost = conn.execute(
            """
            SELECT COALESCE(SUM(i.quantity_on_hand * p.cost), 0) AS total,
                   COALESCE(SUM(i.quantity_on_hand * p.price), 0) AS retail
            FROM inventory i JOIN products p ON p.id = i.product_id
            """
        ).fetchone()

        gross_profit = revenue["total"] - cogs["total"]
        gross_margin = round(gross_profit / revenue["total"] * 100, 1) if revenue["total"] else 0

        # --- AR Aging ---
        ar_aging = conn.execute(
            """
            SELECT
                CASE
                    WHEN julianday('now') - julianday(due_date) <= 0 THEN 'current'
                    WHEN julianday('now') - julianday(due_date) <= 30 THEN '1-30'
                    WHEN julianday('now') - julianday(due_date) <= 60 THEN '31-60'
                    WHEN julianday('now') - julianday(due_date) <= 90 THEN '61-90'
                    ELSE '90+'
                END AS bucket,
                COUNT(*) AS invoice_count,
                COALESCE(SUM(total - amount_paid), 0) AS balance
            FROM invoices
            WHERE status IN ('sent', 'partial', 'overdue')
              AND (total - amount_paid) > 0
            GROUP BY bucket
            ORDER BY CASE bucket
                WHEN 'current' THEN 1 WHEN '1-30' THEN 2 WHEN '31-60' THEN 3
                WHEN '61-90' THEN 4 ELSE 5 END
            """
        ).fetchall()

        ar_detail = conn.execute(
            """
            SELECT invoice_number, c.name AS customer_name, invoice_date, due_date,
                   total, amount_paid, (total - amount_paid) AS balance, status,
                   CAST(julianday('now') - julianday(due_date) AS INTEGER) AS days_overdue
            FROM invoices inv
            JOIN customers c ON c.id = inv.customer_id
            WHERE inv.status IN ('sent', 'partial', 'overdue')
              AND (inv.total - inv.amount_paid) > 0
            ORDER BY due_date ASC
            """
        ).fetchall()

        # --- AP / Outstanding POs ---
        ap_summary = conn.execute(
            """
            SELECT s.name AS supplier_name, s.type,
                   COUNT(po.id) AS open_pos,
                   COALESCE(SUM(po.total), 0) AS open_value
            FROM purchase_orders po
            JOIN suppliers s ON s.id = po.supplier_id
            WHERE po.status IN ('sent', 'partial')
            GROUP BY s.id ORDER BY open_value DESC
            """
        ).fetchall()

        # --- Monthly revenue trend ---
        monthly_revenue = conn.execute(
            f"""
            SELECT strftime('%Y-%m', invoice_date) AS month,
                   COUNT(*) AS invoices,
                   COALESCE(SUM(total), 0) AS revenue,
                   COALESCE(SUM(amount_paid), 0) AS collected
            FROM invoices
            WHERE status NOT IN ('void', 'draft') {inv_where.replace('inv.', '')}
            GROUP BY month ORDER BY month DESC LIMIT 12
            """,
            inv_params,
        ).fetchall()

        # --- Sales by customer ---
        sales_by_customer = conn.execute(
            f"""
            SELECT c.name, c.type,
                   COUNT(inv.id) AS invoice_count,
                   COALESCE(SUM(inv.total), 0) AS total_sales,
                   COALESCE(SUM(inv.amount_paid), 0) AS collected,
                   COALESCE(SUM(inv.total - inv.amount_paid), 0) AS outstanding
            FROM customers c
            LEFT JOIN invoices inv ON inv.customer_id = c.id
                AND inv.status NOT IN ('void', 'draft') {inv_where}
            GROUP BY c.id ORDER BY total_sales DESC
            """,
            inv_params,
        ).fetchall()

        # --- Sales by category ---
        sales_by_category = conn.execute(
            f"""
            SELECT COALESCE(p.category, 'Other') AS category,
                   COALESCE(SUM(il.line_total), 0) AS revenue,
                   COALESCE(SUM(il.quantity), 0) AS units_sold
            FROM invoice_lines il
            JOIN invoices inv ON inv.id = il.invoice_id
            LEFT JOIN products p ON p.id = il.product_id
            WHERE inv.status NOT IN ('void', 'draft') {inv_where}
            GROUP BY category ORDER BY revenue DESC
            """,
            inv_params,
        ).fetchall()

        # --- Top products ---
        top_products = conn.execute(
            f"""
            SELECT COALESCE(p.sku, '—') AS sku,
                   il.description AS name,
                   COALESCE(p.category, 'Other') AS category,
                   COALESCE(SUM(il.quantity), 0) AS qty_sold,
                   COALESCE(SUM(il.line_total), 0) AS revenue
            FROM invoice_lines il
            JOIN invoices inv ON inv.id = il.invoice_id
            LEFT JOIN products p ON p.id = il.product_id
            WHERE inv.status NOT IN ('void', 'draft') {inv_where}
            GROUP BY il.product_id, il.description
            ORDER BY revenue DESC LIMIT 15
            """,
            inv_params,
        ).fetchall()

        # --- Quote pipeline ---
        quote_pipeline = conn.execute(
            f"""
            SELECT status, COUNT(*) AS count, COALESCE(SUM(total), 0) AS value
            FROM quotes q WHERE 1=1 {qt_where.replace('q.', '')}
            GROUP BY status
            """,
            qt_params,
        ).fetchall()

        quote_conversion = conn.execute(
            f"""
            SELECT
                (SELECT COUNT(*) FROM quotes WHERE status = 'accepted' {qt_where.replace('q.', '')}) AS accepted,
                (SELECT COUNT(*) FROM quotes WHERE status IN ('sent','accepted','declined') {qt_where.replace('q.', '')}) AS decided,
                (SELECT COUNT(*) FROM quotes WHERE 1=1 {qt_where.replace('q.', '')}) AS total
            """,
            qt_params * 3,
        ).fetchone()

        # --- Sales by job ---
        sales_by_job = conn.execute(
            f"""
            SELECT j.job_number, j.name AS job_name, c.name AS customer_name, j.status,
                   j.estimated_value,
                   COALESCE(SUM(inv.total), 0) AS invoiced,
                   COALESCE(SUM(inv.amount_paid), 0) AS collected
            FROM jobs j
            JOIN customers c ON c.id = j.customer_id
            LEFT JOIN invoices inv ON inv.job_id = j.id
                AND inv.status NOT IN ('void', 'draft') {inv_where}
            GROUP BY j.id ORDER BY invoiced DESC
            """,
            inv_params,
        ).fetchall()

        # --- Purchases by supplier ---
        purchases_by_supplier = conn.execute(
            f"""
            SELECT s.name, s.type,
                   COUNT(po.id) AS po_count,
                   COALESCE(SUM(po.total), 0) AS total_purchases
            FROM suppliers s
            LEFT JOIN purchase_orders po ON po.supplier_id = s.id
                AND po.status != 'cancelled' {po_where}
            GROUP BY s.id ORDER BY total_purchases DESC
            """,
            po_params,
        ).fetchall()

        # --- PO status ---
        po_status = conn.execute(
            f"""
            SELECT status, COUNT(*) AS count, COALESCE(SUM(total), 0) AS value
            FROM purchase_orders po WHERE 1=1 {po_where.replace('po.', '')}
            GROUP BY status
            """,
            po_params,
        ).fetchall()

        # --- Purchases by category ---
        purchases_by_category = conn.execute(
            f"""
            SELECT p.category,
                   COALESCE(SUM(pol.quantity_ordered), 0) AS units,
                   COALESCE(SUM(pol.line_total), 0) AS spend
            FROM purchase_order_lines pol
            JOIN purchase_orders po ON po.id = pol.purchase_order_id
            JOIN products p ON p.id = pol.product_id
            WHERE po.status != 'cancelled' {po_where}
            GROUP BY p.category ORDER BY spend DESC
            """,
            po_params,
        ).fetchall()

        # --- Inventory ---
        category_inventory = conn.execute(
            """
            SELECT p.category, COUNT(*) AS product_count,
                   SUM(i.quantity_on_hand) AS total_units,
                   SUM(i.quantity_on_hand * p.cost) AS cost_value,
                   SUM(i.quantity_on_hand * p.price) AS retail_value
            FROM products p JOIN inventory i ON i.product_id = p.id
            GROUP BY p.category ORDER BY cost_value DESC
            """
        ).fetchall()

        low_stock = conn.execute(
            """
            SELECT p.sku, p.name, p.category, i.quantity_on_hand, p.reorder_level,
                   s.name AS supplier_name, p.cost,
                   (p.reorder_level - i.quantity_on_hand) AS qty_needed
            FROM inventory i
            JOIN products p ON p.id = i.product_id
            LEFT JOIN suppliers s ON s.id = p.supplier_id
            WHERE i.quantity_on_hand <= p.reorder_level
            ORDER BY qty_needed DESC
            """
        ).fetchall()

        margin = conn.execute(
            """
            SELECT p.sku, p.name, p.category, p.cost, p.price,
                   (p.price - p.cost) AS unit_margin,
                   ROUND((p.price - p.cost) / NULLIF(p.price, 0) * 100, 1) AS margin_pct,
                   i.quantity_on_hand,
                   ROUND(i.quantity_on_hand * (p.price - p.cost), 2) AS potential_profit
            FROM products p
            JOIN inventory i ON i.product_id = p.id
            ORDER BY margin_pct DESC
            """
        ).fetchall()

        # --- Order fulfillment ---
        fulfillment = conn.execute(
            f"""
            SELECT so.order_number, c.name AS customer_name, so.status,
                   so.order_date, so.required_date, so.total,
                   COALESCE(SUM(sol.quantity_ordered), 0) AS qty_ordered,
                   COALESCE(SUM(sol.quantity_shipped), 0) AS qty_shipped
            FROM sales_orders so
            JOIN customers c ON c.id = so.customer_id
            LEFT JOIN sales_order_lines sol ON sol.sales_order_id = so.id
            WHERE so.status NOT IN ('delivered', 'cancelled') {so_where}
            GROUP BY so.id ORDER BY so.required_date ASC
            """,
            so_params,
        ).fetchall()

        # --- Job summary ---
        job_summary = conn.execute(
            """
            SELECT status, COUNT(*) AS count,
                   COALESCE(SUM(estimated_value), 0) AS pipeline_value
            FROM jobs GROUP BY status
            """
        ).fetchall()

    conv_rate = 0
    if quote_conversion["decided"]:
        conv_rate = round(quote_conversion["accepted"] / quote_conversion["decided"] * 100, 1)

    return {
        "filters": {"date_from": date_from, "date_to": date_to},
        "financial": {
            "revenue": revenue["total"],
            "collected": revenue["collected"],
            "outstanding_ar": revenue["outstanding"],
            "invoice_count": revenue["invoice_count"],
            "cogs": cogs["total"],
            "gross_profit": gross_profit,
            "gross_margin_pct": gross_margin,
            "inventory_cost": inventory_cost["total"],
            "inventory_retail": inventory_cost["retail"],
            "open_ap": sum(r["open_value"] for r in ap_summary),
            "ar_aging": rows_to_list(ar_aging),
            "ar_detail": rows_to_list(ar_detail),
            "ap_summary": rows_to_list(ap_summary),
            "monthly_revenue": rows_to_list(monthly_revenue),
        },
        "sales": {
            "by_customer": rows_to_list(sales_by_customer),
            "by_category": rows_to_list(sales_by_category),
            "top_products": rows_to_list(top_products),
            "by_job": rows_to_list(sales_by_job),
            "quote_pipeline": rows_to_list(quote_pipeline),
            "quote_conversion_rate": conv_rate,
            "quote_total": quote_conversion["total"],
        },
        "purchasing": {
            "by_supplier": rows_to_list(purchases_by_supplier),
            "po_status": rows_to_list(po_status),
            "by_category": rows_to_list(purchases_by_category),
        },
        "inventory": {
            "by_category": rows_to_list(category_inventory),
            "low_stock": rows_to_list(low_stock),
            "margins": rows_to_list(margin),
        },
        "operations": {
            "fulfillment": rows_to_list(fulfillment),
            "job_summary": rows_to_list(job_summary),
        },
    }

File: static/js/reports.js
/** Full reporting package UI */

let reportData = null;
let reportSection = 'financial';
let reportDateFrom = '';
let reportDateTo = '';

function exportCsv(type) {
  const params = new URLSearchParams({ type });
  if (reportDateFrom) params.set('from', reportDateFrom);
  if (reportDateTo) params.set('to', reportDateTo);
  window.open(`/api/reports/export?${params}`, '_blank');
}

function reportTable(title, headers, rows, renderRow, exportType) {
  return `
    <div class="card report-panel">
      <div class="card-header">
        <div class="card-title">${title}</div>
        ${exportType ? `<button class="btn btn-sm btn-secondary export-btn" data-export="${exportType}">⬇ CSV</button>` : ''}
      </div>
      <div class="table-wrap"><table>
        <thead><tr>${headers.map(h => `<th>${h}</th>`).join('')}</tr></thead>
        <tbody>${rows.length ? rows.map(renderRow).join('') : `<tr><td colspan="${headers.length}" class="text-muted">No data for selected period</td></tr>`}</tbody>
      </table></div>
    </div>`;
}

function barChart(items, valueKey, labelKey, maxBars = 8) {
  const top = items.slice(0, maxBars);
  const max = Math.max(...top.map(i => i[valueKey] || 0), 1);
  const moneyKeys = ['total', 'revenue', 'spend', 'value', 'balance', 'collected', 'invoiced', 'purchases', 'sales'];
  return `<div class="bar-chart">${top.map(item => {
    const val = item[valueKey] || 0;
    const pct = Math.round((val / max) * 100);
    const display = typeof val === 'number' && moneyKeys.some(k => valueKey.includes(k)) ? fmt(val) : val;
    return `<div class="bar-row">
      <div class="bar-label">${item[labelKey]}</div>
      <div class="bar-track"><div class="bar-fill" style="width:${pct}%"></div></div>
      <div class="bar-value">${display}</div>
    </div>`;
  }).join('')}</div>`;
}

function renderFinancialReports(f) {
  return `
    <div class="metric-grid mb-20">
      <div class="metric-card"><div class="metric-label">Total Revenue</div><div class="metric-value gold">${fmt(f.revenue)}</div></div>
      <div class="metric-card"><div class="metric-label">Collected</div><div class="metric-value green">${fmt(f.collected)}</div></div>
      <div class="metric-card"><div class="metric-label">Outstanding A/R</div><div class="metric-value red">${fmt(f.outstanding_ar)}</div></div>
      <div class="metric-card"><div class="metric-label">COGS (Received POs)</div><div class="metric-value blue">${fmt(f.cogs)}</div></div>
      <div class="metric-card"><div class="metric-label">Gross Profit</div><div class="metric-value green">${fmt(f.gross_profit)}</div></div>
      <div class="metric-card"><div class="metric-label">Gross Margin</div><div class="metric-value purple">${f.gross_margin_pct}%</div></div>
      <div class="metric-card"><div class="metric-label">Inventory (Cost)</div><div class="metric-value blue">${fmt(f.inventory_cost)}</div></div>
      <div class="metric-card"><div class="metric-label">Open A/P (POs)</div><div class="metric-value amber">${fmt(f.open_ap)}</div></div>
    </div>

    <div class="grid grid-2 mb-20">
      ${reportTable('Profit & Loss Summary', ['Metric', 'Amount'], [
        { metric: 'Revenue', amount: f.revenue },
        { metric: 'Cost of Goods Sold', amount: f.cogs },
        { metric: 'Gross Profit', amount: f.gross_profit },
        { metric: 'Gross Margin', amount: `${f.gross_margin_pct}%` },
        { metric: 'Inventory at Cost', amount: f.inventory_cost },
        { metric: 'Inventory at Retail', amount: f.inventory_retail },
      ], r => `<tr><td>${r.metric}</td><td><strong>${typeof r.amount === 'number' ? fmt(r.amount) : r.amount}</strong></td></tr>`)}

      ${reportTable('A/R Aging Summary', ['Bucket', 'Invoices', 'Balance'], f.ar_aging,
        r => `<tr><td>${r.bucket}</td><td>${r.invoice_count}</td><td class="text-accent">${fmt(r.balance)}</td></tr>`)}
    </div>

    ${reportTable('A/R Detail — Open Invoices', ['Invoice', 'Customer', 'Due', 'Balance', 'Days Overdue', 'Status'],
      f.ar_detail, r => `<tr>
        <td><strong>${r.invoice_number}</strong></td><td>${r.customer_name}</td>
        <td>${fmtDate(r.due_date)}</td><td class="text-accent">${fmt(r.balance)}</td>
        <td class="${r.days_overdue > 0 ? 'low-stock' : ''}">${r.days_overdue > 0 ? r.days_overdue : '—'}</td>
        <td>${badge(r.status)}</td></tr>`, 'ar_aging')}

    <div class="grid grid-2 mt-16">
      ${reportTable('Open A/P by Supplier', ['Supplier', 'Type', 'Open POs', 'Value'], f.ap_summary,
        r => `<tr><td>${r.supplier_name}</td><td>${badge(r.type)}</td><td>${r.open_pos}</td><td>${fmt(r.open_value)}</td></tr>`, 'ap_summary')}

      ${reportTable('Monthly Revenue Trend', ['Month', 'Invoices', 'Revenue', 'Collected'], f.monthly_revenue,
        r => `<tr><td>${r.month}</td><td>${r.invoices}</td><td>${fmt(r.revenue)}</td><td class="text-success">${fmt(r.collected)}</td></tr>`, 'monthly_revenue')}
    </div>
  `;
}

function renderSalesReports(s) {
  return `
    <div class="metric-grid mb-20" style="grid-template-columns: repeat(3, 1fr);">
      <div class="metric-card"><div class="metric-label">Quote Conversion Rate</div><div class="metric-value green">${s.quote_conversion_rate}%</div></div>
      <div class="metric-card"><div class="metric-label">Total Quotes</div><div class="metric-value blue">${s.quote_total}</div></div>
      <div class="metric-card"><div class="metric-label">Quote Pipeline Value</div><div class="metric-value gold">${fmt(s.quote_pipeline.reduce((a, q) => a + q.value, 0))}</div></div>
    </div>

    <div class="grid grid-2 mb-20">
      <div class="card report-panel">
        <div class="card-header"><div class="card-title">Revenue by Category</div></div>
        ${barChart(s.by_category, 'revenue', 'category')}
      </div>
      <div class="card report-panel">
        <div class="card-header"><div class="card-title">Quote Pipeline</div></div>
        ${s.quote_pipeline.length ? s.quote_pipeline.map(q => `
          <div class="pipeline-row"><span>${badge(q.status)}</span><span>${q.count} quotes</span><span class="text-accent">${fmt(q.value)}</span></div>
        `).join('') : '<p class="text-muted">No quotes in period</p>'}
      </div>
    </div>

    ${reportTable('Sales by Customer', ['Customer', 'Type', 'Invoices', 'Sales', 'Collected', 'Outstanding'],
      s.by_customer, r => `<tr><td><strong>${r.name}</strong></td><td>${badge(r.type)}</td><td>${r.invoice_count}</td>
      <td>${fmt(r.total_sales)}</td><td class="text-success">${fmt(r.collected)}</td><td class="text-accent">${fmt(r.outstanding)}</td></tr>`, 'sales_by_customer')}

    <div class="grid grid-2 mt-16">
      ${reportTable('Top Products by Revenue', ['SKU', 'Product', 'Category', 'Qty', 'Revenue'], s.top_products,
        r => `<tr><td><code>${r.sku}</code></td><td>${r.name}</td><td>${r.category}</td><td>${r.qty_sold}</td><td>${fmt(r.revenue)}</td></tr>`, 'top_products')}

      ${reportTable('Sales by Category', ['Category', 'Units', 'Revenue'], s.by_category,
        r => `<tr><td>${r.category}</td><td>${r.units_sold}</td><td>${fmt(r.revenue)}</td></tr>`, 'sales_by_category')}
    </div>

    ${reportTable('Sales by Job', ['Job #', 'Job', 'Customer', 'Status', 'Estimated', 'Invoiced', 'Collected'],
      s.by_job, r => `<tr><td><strong>${r.job_number}</strong></td><td>${r.job_name}</td><td>${r.customer_name}</td>
      <td>${badge(r.status)}</td><td>${fmt(r.estimated_value)}</td><td>${fmt(r.invoiced)}</td><td class="text-success">${fmt(r.collected)}</td></tr>`, 'sales_by_job')}
  `;
}

function renderPurchasingReports(p) {
  const totalSpend = p.by_supplier.reduce((a, s) => a + s.total_purchases, 0);
  return `
    <div class="metric-grid mb-20" style="grid-template-columns: repeat(3, 1fr);">
      <div class="metric-card"><div class="metric-label">Total Purchases</div><div class="metric-value gold">${fmt(totalSpend)}</div></div>
      <div class="metric-card"><div class="metric-label">Active Suppliers</div><div class="metric-value blue">${p.by_supplier.filter(s => s.po_count > 0).length}</div></div>
      <div class="metric-card"><div class="metric-label">PO Status Types</div><div class="metric-value purple">${p.po_status.length}</div></div>
    </div>

    <div class="grid grid-2 mb-20">
      <div class="card report-panel">
        <div class="card-header"><div class="card-title">Spend by Supplier</div>
          <button class="btn btn-sm btn-secondary export-btn" data-export="purchases_by_supplier">⬇ CSV</button></div>
        ${barChart(p.by_supplier.filter(s => s.total_purchases > 0), 'total_purchases', 'name')}
      </div>
      <div class="card report-panel">
        <div class="card-header"><div class="card-title">PO Status Breakdown</div></div>
        ${p.po_status.map(s => `<div class="pipeline-row"><span>${badge(s.status)}</span><span>${s.count} orders</span><span class="text-accent">${fmt(s.value)}</span></div>`).join('') || '<p class="text-muted">No POs in period</p>'}
      </div>
    </div>

    ${reportTable('Purchases by Supplier', ['Supplier', 'Type', 'PO Count', 'Total Spend'], p.by_supplier,
      r => `<tr><td><strong>${r.name}</strong></td><td>${badge(r.type)}</td><td>${r.po_count}</td><td>${fmt(r.total_purchases)}</td></tr>`, 'purchases_by_supplier')}

    ${reportTable('Purchases by Category', ['Category', 'Units Ordered', 'Spend'], p.by_category,
      r => `<tr><td>${r.category}</td><td>${r.units}</td><td>${fmt(r.spend)}</td></tr>`, 'purchases_by_category')}
  `;
}

function renderInventoryReports(inv) {
  const totalCost = inv.by_category.reduce((a, c) => a + c.cost_value, 0);
  const totalRetail = inv.by_category.reduce((a, c) => a + c.retail_value, 0);
  return `
    <div class="metric-grid mb-20">
      <div class="metric-card"><div class="metric-label">Inventory at Cost</div><div class="metric-value gold">${fmt(totalCost)}</div></div>
      <div class="metric-card"><div class="metric-label">Inventory at Retail</div><div class="metric-value green">${fmt(totalRetail)}</div></div>
      <div class="metric-card"><div class="metric-label">Markup Potential</div><div class="metric-value purple">${fmt(totalRetail - totalCost)}</div></div>
      <div class="metric-card"><div class="metric-label">Low Stock SKUs</div><div class="metric-value red">${inv.low_stock.length}</div></div>
    </div>

    ${reportTable('Inventory Valuation by Category', ['Category', 'Products', 'Units', 'Cost Value', 'Retail Value'],
      inv.by_category, r => `<tr><td>${r.category}</td><td>${r.product_count}</td><td>${r.total_units}</td>
      <td>${fmt(r.cost_value)}</td><td class="text-success">${fmt(r.retail_value)}</td></tr>`, 'inventory_valuation')}

    <div class="grid grid-2 mt-16">
      ${reportTable('Low Stock / Reorder Report', ['SKU', 'Product', 'On Hand', 'Reorder', 'Need', 'Supplier', 'Cost'],
        inv.low_stock, r => `<tr><td><code>${r.sku}</code></td><td>${r.name}</td>
        <td class="low-stock">${r.quantity_on_hand}</td><td>${r.reorder_level}</td><td>${r.qty_needed}</td>
        <td>${r.supplier_name || '—'}</td><td>${fmt(r.cost)}</td></tr>`, 'low_stock')}

      ${reportTable('Product Margin Analysis', ['SKU', 'Product', 'Cost', 'Price', 'Margin %', 'On Hand', 'Potential Profit'],
        inv.margins, r => `<tr><td><code>${r.sku}</code></td><td>${r.name}</td><td>${fmt(r.cost)}</td><td>${fmt(r.price)}</td>
        <td class="text-success">${r.margin_pct}%</td><td>${r.quantity_on_hand}</td><td>${fmt(r.potential_profit)}</td></tr>`, 'product_margins')}
    </div>
  `;
}

function renderOperationsReports(op) {
  const activeJobs = op.job_summary.find(j => j.status === 'active');
  return `
    <div class="metric-grid mb-20" style="grid-template-columns: repeat(3, 1fr);">
      <div class="metric-card"><div class="metric-label">Open Orders to Fulfill</div><div class="metric-value blue">${op.fulfillment.length}</div></div>
      <div class="metric-card"><div class="metric-label">Active Jobs</div><div class="metric-value green">${activeJobs?.count || 0}</div></div>
      <div class="metric-card"><div class="metric-label">Active Pipeline Value</div><div class="metric-value gold">${fmt(activeJobs?.pipeline_value || 0)}</div></div>
    </div>

    <div class="grid grid-2 mb-20">
      <div class="card report-panel">
        <div class="card-header"><div class="card-title">Job Status Summary</div></div>
        ${op.job_summary.map(j => `<div class="pipeline-row"><span>${badge(j.status)}</span><span>${j.count} jobs</span><span class="text-accent">${fmt(j.pipeline_value)}</span></div>`).join('')}
      </div>
    </div>

    ${reportTable('Order Fulfillment — Open Sales Orders', ['Order #', 'Customer', 'Status', 'Required', 'Total', 'Ordered', 'Shipped'],
      op.fulfillment, r => `<tr><td><strong>${r.order_number}</strong></td><td>${r.customer_name}</td><td>${badge(r.status)}</td>
      <td>${fmtDate(r.required_date)}</td><td>${fmt(r.total)}</td><td>${r.qty_ordered}</td><td>${r.qty_shipped}</td></tr>`, 'order_fulfillment')}
  `;
}

function renderReportSection() {
  if (!reportData) return;
  const sections = {
    financial: renderFinancialReports(reportData.financial),
    sales: renderSalesReports(reportData.sales),
    purchasing: renderPurchasingReports(reportData.purchasing),
    inventory: renderInventoryReports(reportData.inventory),
    operations: renderOperationsReports(reportData.operations),
  };
  $('#report-body').innerHTML = sections[reportSection] || '';
  document.querySelectorAll('.export-btn').forEach(btn => {
    btn.onclick = () => exportCsv(btn.dataset.export);
  });
}

export async function renderReports(deps) {
  const { topbarActions, content, api, fmt, fmtDate, badge, $ } = deps;
  Object.assign(window, { fmt, fmtDate, badge, $ });

  topbarActions().innerHTML = `
    <button class="btn btn-secondary" id="print-reports">🖨 Print</button>
    <button class="btn btn-primary" id="refresh-reports">↻ Refresh</button>
  `;

  content().innerHTML = `
    <div class="report-toolbar">
      <div class="report-filters">
        <label>From <input type="date" id="report-from" value="${reportDateFrom}"></label>
        <label>To <input type="date" id="report-to" value="${reportDateTo}"></label>
        <button class="btn btn-sm btn-secondary" id="apply-report-filter">Apply</button>
        <button class="btn btn-sm btn-secondary" id="clear-report-filter">Clear</button>
      </div>
    </div>
    <nav class="report-tabs" id="report-tabs">
      <button class="report-tab ${reportSection === 'financial' ? 'active' : ''}" data-section="financial">💰 Financial</button>
      <button class="report-tab ${reportSection === 'sales' ? 'active' : ''}" data-section="sales">📊 Sales</button>
      <button class="report-tab ${reportSection === 'purchasing' ? 'active' : ''}" data-section="purchasing">🛒 Purchasing</button>
      <button class="report-tab ${reportSection === 'inventory' ? 'active' : ''}" data-section="inventory">📦 Inventory</button>
      <button class="report-tab ${reportSection === 'operations' ? 'active' : ''}" data-section="operations">⚙ Operations</button>
    </nav>
    <div id="report-body"><div class="text-muted">Loading reports...</div></div>
  `;

  async function loadReports() {
    $('#report-body').innerHTML = '<div class="text-muted">Loading reports...</div>';
    const params = new URLSearchParams();
    if (reportDateFrom) params.set('from', reportDateFrom);
    if (reportDateTo) params.set('to', reportDateTo);
    const qs = params.toString();
    reportData = await api(`reports/full${qs ? '?' + qs : ''}`);
    renderReportSection();
  }

  $('#report-tabs').querySelectorAll('.report-tab').forEach(tab => {
    tab.onclick = () => {
      reportSection = tab.dataset.section;
      $('#report-tabs').querySelectorAll('.report-tab').forEach(t => t.classList.toggle('active', t.dataset.section === reportSection));
      renderReportSection();
    };
  });

  $('#apply-report-filter').onclick = () => {
    reportDateFrom = $('#report-from').value;
    reportDateTo = $('#report-to').value;
    loadReports();
  };
  $('#clear-report-filter').onclick = () => {
    reportDateFrom = reportDateTo = '';
    $('#report-from').value = $('#report-to').value = '';
    loadReports();
  };
  $('#refresh-reports').onclick = loadReports;
  $('#print-reports').onclick = () => window.print();

  await loadReports();
}
